Add specification for ReferenceHandling (#1)
Jozkee/corefx: src/System.Text.Json/docs/ReferenceHandling_spec.md

> # Motivation

Reviewer: Maybe add a background on terminology and what the problem space has.

Something like this:

We have the following concepts:
1) Reference loops: If the .NET object graph being serialized contains a cycle (the runtime instance).
2) Preserving duplicate references: If the .NET object graph contains properties with duplicate references (at runtime).
3) Emitting and parsing metadata: Emitting more information to the JSON (with potentially changing the schema) to enable reference preservation when round-tripping.

> Currently there is no mechanism to prevent infinite looping in circular objects nor to preserve references when using System.Text.Json.

Reviewer: ```suggestion
Currently there is no mechanism to prevent infinite looping in circular objects (while serializing) nor to preserve references that round-trip when using System.Text.Json.
```

> Community is heavily requesting this feature since is consider by many as a very common scenario, specially when serializing POCOs that came from an ORM Framework, such as Entity Framework; even though JSON specifiacation does not support reference loops by default. Therefore this will be shipped as an opt-in feature.

Reviewer: ```suggestion
This is a heavily requested feature since it is consider by many as a very common scenario, specially when serializing POCOs that came from an ORM Framework, such as Entity Framework; even though the JSON specification does not support reference loops by default. Therefore, this will be implemented as an opt-in feature (for both serialization and deserialization).
```

> This is a heavily requested feature since it is consider by many as a very common scenario, specially when serializing POCOs that came from an ORM Framework, such as Entity Framework; even though the JSON specification does not support reference loops by default. Therefore, this will be implemented as an opt-in feature (for both serialization and deserialization).

Reviewer: ```suggestion
This is a heavily requested feature since it is considered by many as a very common scenario, specially when serializing POCOs that came from an ORM Framework, such as Entity Framework; even though the JSON specification does not support reference loops by default. Therefore, this will be implemented as an opt-in feature (for both serialization and deserialization).
```

> The current solution to deal with reference loops is to rely in MaxDepth and throw a JsonException after it is exceeded. Now, this is a decent and cheap solution but we will also offer other not-so-cheap options to deal with this problem while keeping the current one in order to not affect the out-of-the-box performance.

Reviewer: ```suggestion
The current solution to deal with cycles in the object graph while serializing is to rely on `MaxDepth` and throw a `JsonException` after it is exceeded. This was done to avoid perf overhead for cycle detection in the common case. The goal is to enable the new opt-in feature with minimal impact to existing performance.
```

Reviewer: ```text
The current solution to deal with cycles in the object graph while serializing is to rely on `MaxDepth` and throw a `JsonException` after it is exceeded. This was done to avoid perf overhead for cycle detection in the common case. The goal is to enable the new opt-in feature with minimal impact to existing performance.
```

Reviewer 2: Where did those HTML tags came from?

Reviewer: Don't know. Something broke with the suggestions. That's why I copied the suggestion as regular text.

> # Prior art

Reviewer: What about what other languages/libraries do? I would add that here too.

Reviewer: Given we are basically talking about what other libraries do, when you add more info here, consider wording the title as "Other libraries" like you had before.

> The current solution to deal with reference loops is to rely in MaxDepth and throw a JsonException after it is exceeded. Now, this is a decent and cheap solution but we will also offer other not-so-cheap options to deal with this problem while keeping the current one in order to not affect the out-of-the-box performance.

Reviewer: ```suggestion
The current solution to deal with cycles in the object graph while serializing is to rely on `MaxDepth` and throw a `JsonException` after it is exceeded. This was done to avoid perf overhead for cycle detection in the common case. The goal is to enable the new opt-in feature with minimal impact to existing performance.

```

> Json.Net contains settings that you can enable to deal with such problems.

Reviewer: Please use `Newtonsoft.Json` everywhere rather than `Json.NET`. Json.Net is confusing to readers (they tend to think that's the built-in library).

* For Serialization:
  * [`ReferenceLoopHandling`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_ReferenceLoopHandling.htm)
  * [`PreserveReferencesHandling`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_PreserveReferencesHandling.htm)
* For Deserialization:
  * [`MetadataPropertyHandling`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_MetadataPropertyHandling.htm)

> When using `ReferenceLoopHandling.Ignore`, other objects that were already seen on the current graph branch will be ignored on Serialization.

Reviewer: ```suggestion
When using `ReferenceLoopHandling.Ignore`, other objects that were already seen on the current graph branch will be ignored on serialization.
```

When using `PreserveReferencesHandling.All` you are signaling that your resulting JSON will contain *metadata* properties `$ref`, `$id` and `$values` which are going to act as reference identifiers (`$id`) and pointers (`$ref`). 

> Now, to read back those references, you have to use `MetadataPropertyHandling.Default` to indicate that *metadata* is expected in the payload passed to the Deserialize method.

Reviewer: ```suggestion
Now, to read back those references, you have to use `MetadataPropertyHandling.Default` to indicate that *metadata* is expected in the payload passed to the `Deserialize` method.
```

# Proposal

> 

Reviewer: I liked the clear description/documentation of the 3 values (assuming we also add ignore) from https://github.com/dotnet/corefx/issues/41002.

Let's add this here too:
> ## In depth
> * **Default**:
>   
>   * **On Serialize**: Throw a JsonException when MaxDepth is exceeded, this may occur by either a Reference Loop or by passing a very deep object. This option will not affect the performance of the serializer.
>   * **On Deserialize**: No effect.
> * **Ignore**:
>   
>   * **On Serialize**: Ignores (skips writing) the property/element where the reference loop is detected.
>   * **On Deserialize**: No effect.
> * **Preserve**:
>   
>   * **On Serialize**: When writing complex types, the serializer also writes them metadata ($id, $values and $ref) properties in order re-use them by writing a reference to the object or array.
>   * **On Deserialize**: While the other options show no effect on Deserialization, `Preserve` does affect its behavior with the following: Metadata will be expected (although is not mandatory) and the deserializer will try to understand it.

```cs
namespace System.Text.Json.Serialization
{
    /// <summary>
    /// This class defines the various ways the <see cref="JsonSerializer"/> can deal with references on Serialization and Deserialization.
    /// </summary>
    public sealed class ReferenceHandling
    {
        //built-in options.
        public static ReferenceHandling Default => new ReferenceHandling(PreserveReferencesHandling.None);
        public static ReferenceHandling Preserve => new ReferenceHandling(PreserveReferencesHandling.All);

>         // For future, change to public if needed.

Reviewer: I would more clearly separate the actual proposal and types and what we might do in the future. Also, remove the implementation details. Right now this looks too cluttered, making it harder to reason about for new comers.

Reviewer 2: What if we only expose the built-in static properties, since you will not get a different result by using the public constructor?

Reviewer: That makes sense to me. We should still show the internal types/expected future APIs (preferably separate from the API surface) so we can discuss naming and the feature all up.

        internal PreserveReferencesHandling PreserveHandlingOnSerialize => _preserveHandlingOnSerialize;
        internal PreserveReferencesHandling PreserveHandlingOnDeserialize => _preserveHandlingOnDeserialize;
        internal ReferenceLoopHandling LoopHandling => _loopHandling;

        private PreserveReferencesHandling _preserveHandlingOnSerialize;
        private PreserveReferencesHandling _preserveHandlingOnDeserialize;
        private ReferenceLoopHandling _loopHandling;

        public ReferenceHandling(PreserveReferencesHandling handling) : this(handling, handling, ReferenceLoopHandling.Error) { }

>         // For future, someone may want to define their own custom Handler with different behaviors of PreserveReferenceHandling on Serialize vs Deserialize and another ReferenceLoopHandling, such as ignore if added in a future.

Reviewer: Add line-wraps for the comments/code so it is all visible on the screen

        private ReferenceHandling(PreserveReferencesHandling preserveHandlingOnSerialize, PreserveReferencesHandling preserveHandlingOnDeserialize, ReferenceLoopHandling loopHandling)
        {
            _preserveHandlingOnSerialize = preserveHandlingOnSerialize;
            _preserveHandlingOnDeserialize = preserveHandlingOnDeserialize;
            _loopHandling = loopHandling;
        }
    }

    internal enum ReferenceLoopHandling
    {
        Error = 0,
        // For future if requested by the community.
        // Ignore = 1,
    }

>     public enum PreserveReferencesHandling

Reviewer: Should we name this differently from `Newtonsoft.Json` and maybe avoid the redundant use of "ReferenceHandling"?

    {
        None = 0,
        All = 1,
        // For future if requested by the community.
        // Objects = 2,
        // Arrays = 3,
    }
}

namespace System.Text.Json
{
    public partial class JsonSerializerOptions
    {

>         public ReferenceHandling ReferenceHandling { get; set; } = ReferenceHandling.Default;

Reviewer: Explicitly call out the semantics of the default behavior for both serialize and deserialize (and how it isn't changing).

Reviewer: What happens by default when we deserialize JSON containing metadata properties? Also, is there a way to ignore them?

Reviewer 2: > What happens by default when we deserialize JSON containing metadata properties?

They are treated as any other property, if they are not mapped to a C# property (by using attribute `[JsonPropertyName("$id")]`, for example), they will be added to the `JsonExtensionData` dictionary if there is any.

> Is there a way to ignore them?

If by "ignore them" you mean completely skip them when parsing the JSON; no, there is no way, nor reason to do that. Newtonsoft.Json is also incapable of doing that.

    }
}
```

> For System.Text.Json, the goal is to stick to the same *metadata* semantics for preserve from Json.Net and provide a similar usage in `JsonSerializerOptions` that encompasses at least the needed options for an immediate solution, which is to provide reference preservation.

Reviewer: ```suggestion
For System.Text.Json, the goal is to stick to the same *metadata* semantics for preserve from Json.Net and provide a similar usage in `JsonSerializerOptions` that encompasses the needed options (i.e. provide reference preservation).
```

> This API is about exposing the `ReferenceHandling` property as a class type, in order to make it very extensible for future; and provide built-in static instances of `Default` and `Preserve` that are useful to enable the most common behaviors by just setting those in `JsonSerializerOptions.ReferenceHandling`.

Reviewer: ```suggestion
This API is exposing the `ReferenceHandling` property as a class, to be extensible in the future; and provide built-in static instances of `Default` and `Preserve` that are useful to enable the most common behaviors by just setting those in `JsonSerializerOptions.ReferenceHandling`.
```

> With ReferenceHandling being a class, we can exclude things that, as of now, we are not sure if they need to be shipped, in example, the `ReferenceLoopHandling.Ignore` option or the Object and Array granularity of Json.Net's `PreserveReferencesHandling`.

Reviewer: ```suggestion
With `ReferenceHandling` being a class, we can exclude things that, as of now, we are not sure are required and add them later based on customer feedback. For example, the `Object` and `Array` granularity of `Newtonsoft.Json's` ``PreserveReferencesHandling` or the `ReferenceLoopHandling.Ignore` option.
```

> With `ReferenceHandling` being a class, we can exclude things that, as of now, we are not sure are required and add them later based on customer feedback. For example, the `Object` and `Array` granularity of `Newtonsoft.Json's` ``PreserveReferencesHandling` or the `ReferenceLoopHandling.Ignore` option.

Reviewer: ```suggestion
With `ReferenceHandling` being a class, we can exclude things that, as of now, we are not sure are required and add them later based on customer feedback. For example, the `Object` and `Array` granularity of `Newtonsoft.Json's` `PreserveReferencesHandling` or the `ReferenceLoopHandling.Ignore` option.
```

> In a future, we might also be able to extend the Deserialization behaviors, in order to provide the granularity of Serialization on Deserialization. Or we could also be able to define independant behaviors for both Serialization and Deserialization; i.e: someone may want to emit payloads with preserved references but he does not want to read them.

Reviewer: ```suggestion
In the future, we might also be able to extend the deserialization behaviors, in order to provide the granularity of Serialization on Deserialization. We could also provide ways to customize the serialization and deserialization behavior independently (i.e: someone may want to emit payloads with preserved references but they do not want to read them).
```

> In a future, we might also be able to extend the Deserialization behaviors, in order to provide the granularity of Serialization on Deserialization. Or we could also be able to define independant behaviors for both Serialization and Deserialization; i.e: someone may want to emit payloads with preserved references but he does not want to read them.

Reviewer: I don't understand what you mean by: 
```text
In a future, we might also be able to extend the Deserialization behaviors, in order to provide the granularity of Serialization on Deserialization.
```

Reviewer 2: So, right now we either **write and read** **All** preserved references (`PreserveReferencesHandling.All`) or we don't (`PreserveReferencesHandling.None`). 

If we expose this constructor:
```cs
private ReferenceHandling(PreserveReferencesHandling preserveHandlingOnSerialize, PreserveReferencesHandling preserveHandlingOnDeserialize, ReferenceLoopHandling loopHandling)
```

We can define different granularity for serialization vs deserialization.

Reviewer: > We can define different granularity for serialization vs deserialization.

Isn't that what you are saying in the second sentence? So, what's the first sentence trying to say?

> Or we could also be able to define independant behaviors for both Serialization and Deserialization; i.e: someone may want to emit payloads with preserved references but he does not want to read them.

> In a future, we might also be able to extend the Deserialization behaviors, in order to provide the granularity of Serialization on Deserialization. Or we could also be able to define independant behaviors for both Serialization and Deserialization; i.e: someone may want to emit payloads with preserved references but he does not want to read them.

Reviewer: Add the support matrix here. Along with what the support matrix could look like in the future (also consider including read ahead in your matrix of things we don't support).

Reviewer: https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_MetadataPropertyHandling.htm

Reviewer: > Add the support matrix here.

Just to clarify, referring to something like this:
https://github.com/dotnet/corefx/issues/38579#issuecomment-556923546

> 

Reviewer: Also add a note that we considered enabling this feature via attributes (maybe with sample code that you had, if it worked), but deferred that to the future. If we can't do it for some reason, mention that.

# Examples

Having the following class:
```cs
class Employee 
{ 
    string Name { get; set; }
    Employee Manager { get; set; }
    List<Employee> Subordinates { get; set; }
}
```

> ## Using Preserve on Serialize

Reviewer: Add the sample for Ignore as well, even if we don't end up adding it.

On System.Text.Json:
```cs
public static void WritePreservingReference()
{
    var bob = new Employee { Name = "Bob" };
    var angela = new Employee { Name = "Angela" };

    angela.Manager = bob;
    bob.Subordinates = new List<Employee>{ angela };

    var options = new JsonSerializerOptions
    {
        ReferenceHandling = ReferenceHandling.Preserve
        WriteIndented = true,
    };

    string json = JsonSerializer.Serialize(angela, options);
    Console.Write(json);
}
```

On Newtonsoft's Json.Net:
```cs
public static void WritePreservingReference()
{
    var bob = new Employee { Name = "Bob" };
    var angela = new Employee { Name = "Angela" };

    angela.Manager = bob;
    bob.Subordinates = new List<Employee>{ angela };

    var settings = new JsonSerializerSettings
    {
        PreserveReferencesHandling = PreserveReferencesHandling.All
        Formatting = Formatting.Indented 
    };

    string json = JsonConvert.SerializeObject(angela, settings);
    Console.Write(json);
}
```

Output:
```jsonc
{
    "$id": "1",
    "Name": "Angela",
    "Manager": {
        "$id": "2",
        "Name": "Bob",

>         "Subordinates": { //Note how the Subordinates' square braces are replaced with curly braces in order to include $id and $values properties, $values will now hold whatever value was meant for the Subordinates list.

Reviewer: Add line-wraps for the comments so it is more readable on GitHub.

            "$id": "3",
            "$values": [
                {  //Note how this object denotes reference to Angela that was previously serialized.
                    "$ref": "1"
                }
            ]
        }            
    }
}
```

> ## Using Preserve on Deserialize

Reviewer: Show sample example of what the usage might look like if we add the APIs in vNext (outside of the built-in factory methods).

    {
        ReferenceHandling = ReferenceHandling.Preserve
    };

    Employee angela = JsonSerializer.Deserialize<Employee>(json, options);
    Console.WriteLine(object.ReferenceEquals(angela, angela.Manager.Subordinates[0])); //prints: true.
}
```

On Newtonsoft's Json.Net:
```cs
public static void ReadJsonWithPreservedReferences(){
    string json = 
    @"{
        ""$id"": ""1"",
        ""Name"": ""Angela"",
        ""Manager"": {
            ""$id"": ""2"",
            ""Name"": ""Bob"",
            ""Subordinates"": {
                ""$id"": ""3"",
                ""$values"": [
                    { 
                        ""$ref"": ""1"" 
                    }
                ]
            }            
        }
    }";

    var options = new JsonSerializerSettings
    {
        MetadataPropertyHanding = MetadataPropertyHandling.Default //Json.Net reads metadata by default, just setting the option for ilustrative purposes.
    };

    Employee angela = JsonConvert.DeserializeObject<Employee>(json, settings);
    Console.WriteLine(object.ReferenceEquals(angela, angela.Manager.Subordinates[0])); //prints: true.
}
```

# Ground rules
## Unsupported types

> Basically, any type that uses a `EnumerableConverter` and tries to be preserved will throw.

Reviewer: `EnumerableConverter` is an internal type. Is it necessary to mention this? I would re-state this from the end-user's perspective and also explain why we'd throw here.

Consider moving this down below since it isn't the most important ground rule.

* **Immutable types**: i.e: `ImmutableList` and `ImmutableDictionary`
* **System.Array**

> Aside from above, the Deserializer will throw when a CLR value type is passed as preserved when `ReferenceHandling.Preserve` is set. 

Reviewer: I don't know what you mean by this.

Reviewer 2: We will throw when we read Preserved References and a value type, such as a Struct, contains a $ref.

Reviewer: I see. I would write it like this then:

```suggestion
Aside from above, the deserializer will throw when it reads `$ref` within a property that matches to a value type (like a struct), including when `ReferenceHandling.Preserve` is set.
```

> As a rule of thumb, we should throw on all cases where the payload contains metadata that is impossible to create with the `JsonSerializer`, however, this conflicts with feature parity in Json.Net; those scenarios are described below.

Reviewer: ```suggestion
As a rule of thumb, we throw on all cases where the JSON payload being read contains any metadata that is impossible to create with the `JsonSerializer` (i.e. it was hand modified). However, this conflicts with feature parity in Newtonsoft.Json; those scenarios are described below.
```

Reviewer: This doesn't belong under the `Unsupported types` imo, but rather should go under the `Ground rules` heading, on its own.

## Reference objects ($ref)

* Regular property **before** `$ref`.
  * **Json.Net**: `$ref` is ignored if a regular property is previously found in the object.
  * **S.T.Json**: Throw - Reference objects cannot contain other properties.

```json
{
    "$id": "1",
    "Name": "Angela",
    "ReportsTo": {
        "Name": "Bob",
        "$ref": "1"
    }
}
```

* Regular property **after** `$ref`.
  * **Json.Net**: Throw - Additional content found in JSON reference object.
  * **S.T.Json**: Throw - Reference objects cannot contain other properties.

```json
{
    "$id": "1",
    "Name": "Angela",
    "ReportsTo":{
        "$ref": "1",
        "Name": "Angela" 
    }
}
```
 
* Metadata property **before** `$ref`:
  * **Json.Net**: `$id` is disregarded and the reference is set.
  * **S.T.Json**: Throw - Reference objects cannot contain other properties.
```json
{
    "$id": "1",
    "Name": "Angela",
    "ReportsTo": {
        "$id": "2",
        "$ref": "1"
    }
}
```

> * Metadata property **after** `$ref`:

Reviewer: Add rule about `JsonExtensionData` where metadata won't get added to it ever (even if we ignore it with a setting).

Reviewer 2: I though we agreed that metadata will be added to `JsonExtensionData` if we choose to ignore it. There might be people that depend on such properties existing in it.

Reviewer: > I though we agreed that metadata will be added to `JsonExtensionData` if we choose to ignore it.

Is ignoring metadata the default behavior on deserialization? I am fine with that. Please document it as such.

  * **Json.Net**: Throw with the next message: 'Additional content found in JSON reference object'.
  * **S.T.Json**: Throw - Reference objects cannot contain other properties.
```json
{
    "$id": "1",
    "Name": "Angela",
    "ReportsTo": {
        "$ref": "1",
        "$id": "2"
    }
}
```
 

> * reference object is before preserved object (or preserved object was never spotted):

Reviewer: ```suggestion
* Reference object is before preserved object (or preserved object was never spotted):
```

  * **Json.Net**: Reference object evaluates as `null`.
  * **S.T.Json**: Reference object evaluates as `null`.
```json
[
    {
        "$ref": "1"
    },
    {
        "$id": "1",
        "Name": "Angela"
    }
]
```

> ## Preserved objects ($id)

Reviewer: Are these metadata keywords reserved? What happens if someone puts `JsonPropertyName("$id")` attribute on some time? What about camelCase/snakeCase matching or CaseInsensitive comparison?

* Having more than one `$id` in the same object:
  * **Json.Net**: last one wins, in the example, the reference object evaluates to `null` (if `$ref` would be `"2"`, it would evaluate to itself).
  * **S.T.Json**: Throw - Object already defines a reference identifier.
```json
{
    "$id": "1",
    "$id": "2",
    "Name": "Angela",
    "ReportsTo": {
        "$ref": "1"
    }
}
```

* `$id` is not the first property:
  * **Json.Net**: Object is not preserved and cannot be referenced, therefore any reference to it would evaluate as null.

>   * **S.T.Json**: We can handle the `$id` not being the first property since we store the reference at the moment we spot the property, I don't think we should throw but keep in mind that this is not a normal payload produced by the serializer.

Reviewer: Why are we deviating from our general rule of thumb here? What's the benefit?

Reviewer 2: It might be that something reads our payload and when emitting it back (i.e: JavaScript), the order of the properties change? But then that also might apply to arrays (`$values` being above `$id`). 

I will switch to throw in this case.

Reviewer: We can throw for this case according to our general rule of thumb and then add a note that we can theoretically support this with little change to the implementation, if needed.

    "ReportsTo": {
        "$ref": "1"
    }
}
```

* `$id` is duplicated (not necessarily nested):
  * **Json.Net**: Throws - Error reading object reference '1'- Inner Exception: ArgumentException: A different value already has the Id '1'.
  * **S.T.Json**: Throws - Duplicated id found while preserving reference.
```json
[
    {
        "$id": "1",
        "Name": "Angela"
    },
    {
        "$id": "1",
        "Name": "Bob"
    }
]
```

## Preserved arrays
A regular array is `[ elem1, elem2 ]`.
A preserved array is written in the next format `{ "$id": "1", "$values": [ elem1, elem2 ] }`

* Preserved array does not contain any metadata:
  * **Json.Net**: Throws - Cannot deserialize the current JSON object into type 'System.Collections.Generic.List`1
  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.

  ```json
  {}
  ```

* Preserved array only contains $id:
  * **Json.Net**: Throws - Cannot deserialize the current JSON object into type 'System.Collections.Generic.List`1
  * **S.T.Json**: Throw - Preserved array $values property was not present or its value is not an array.

  ```json
  {
      "$id": "1"
  }
  ```

> * Preserved array only contains $values:

Reviewer: Put the metadata keywords within backticks wherever used:
```suggestion
* Preserved array only contains `$values`:
```

  * **Json.Net**: Does not throw and the payload evaluates to the array in the property.

>   * **S.T.Json**: Throw - Preserved arrays canot lack an identifier.

Reviewer: ```suggestion
  * **S.T.Json**: Throw - Preserved arrays cannot lack an identifier.
```

Reviewer: Consider doing a spell-check across the document (maybe copy/paste the content into MS Word or equivalent to help with that).

  ```json
  {
      "$id": "1",
      "$values": null
  }
  ```

* Preserved array $values property contains value
  * **Json.Net**: Unexpected token while deserializing object: EndObject. Path ''.
  * **S.T.Json**: Throw - The JSON value could not be converted to TArray. Path: $.$values

  ```json
  {
      "$id": "1",
      "$values": 1
  }
  ```

* Preserved array $values property contains object
  * **Json.Net**: Unexpected token while deserializing object: EndObject. Path ''.
  * **S.T.Json**: Throw - The property is already part of a preserved array object, cannot be read as a preserved array.

  ```json
  {
      "$id": "1",
      "$values": {}
  }
  ```

# Notes

1. MaxDepth validation will not be affected by `ReferenceHandling.Preserve`.
2. We are merging the Json.Net types [`ReferenceLoopHandling`]("https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_ReferenceLoopHandling.htm") and [`PreserveReferencesHandling`]("https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_PreserveReferencesHandling.htm") (we are also not including the granularity on this one) into one single class; `ReferenceHandling`.

> 3. While Immutable types and `System.Array`s can be Serialized with Preserve semantics, they will not be supported when trying to Deserialize them as a reference.

Reviewer: Explain why.

> 4. Value types, such as structs that contain preserve semantics, will not be supported when Deserialized as well.

Reviewer: Same here. Why?

> 5. Additional features, such as Converter support, `ReferenceResolver`, `JsonPropertyAttribute.IsReference` and `JsonPropertyAttribute.ReferenceLoopHandling`,  that build on top of `ReferenceLoopHandling` and `PreserveReferencesHandling` were considered but they will not be included in this first effort.

Reviewer: ```suggestion
5. Additional features, such as Converter support, `ReferenceResolver`, `JsonPropertyAttribute.IsReference` and `JsonPropertyAttribute.ReferenceLoopHandling`,  that build on top of `ReferenceLoopHandling` and `PreserveReferencesHandling` were considered but they can be added in the future based on customer requests.
```
